### ISSUE TITLE
Error in interpolation due to absent optional value should result in empty string

### ISSUE THREAD

**Issue by markjschreiber** [chanzuckerberg/miniwdl#523]
When interpolating the following expression

```
Int? length
...
~{"--sjdbOverhang " + (length - 1)}
```

An error is emitted:

```
"error": "IncompatibleOperand",
 "message": "Non-numeric operand to - operator",
```

Although it's a bit vague, this part of the spec suggests that the error in the interpolation should result in the interpolation emitting nothing or an empty string. https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#prepending-a-string-to-an-optional-parameter

**Comment by mlin** [comment]
The written spec applies when the top-level interpolated expression (the entire contents of `~{...}`) to evaluate to null, and also has a [special carve-out](https://github.com/openwdl/wdl/blob/main/versions/1.1/SPEC.md#concatenation-of-optional-values) for string concatenation (`+`) inside an interpolation, allowing nulls to propagate through them. Miniwdl does support those, but I don't believe there's a similar carve-out for *doing math* on optional values as in this example. It's probably a question to take to the OpenWDL spec discussion forum. Is the original workflow open-source somewhere?

**Comment by markjschreiber** [comment]
John Didion on the openwdl slack seems to think the behavior should be to emit an empty string. https://openwdl.slack.com/archives/CT2F743C3/p1636124751003700

I guess the rationale is that the math operation produces an error and therefore the surrounding interpolation expression produces an error and so generates an empty string.

The workflow is one of the GATK best practices RNA SNP workflows:

https://github.com/gatk-workflows/gatk4-rnaseq-germline-snps-indels/blob/master/gatk4-rna-best-practices.wdl#L344

**Comment by mlin** [comment]
I don't see any way to get from the [relevant spec section](https://github.com/openwdl/wdl/blob/main/versions/development/SPEC.md#expression-placeholders-and-string-interpolation) to ignoring any error in an interpolation. Like, would that include syntax errors? @jdidion

> If an expression within a placeholder evaluates to None, then the placeholder is replaced by the empty string.
> Concatenation of Optional Values
> Within expression placeholders the string concatenation operator (+) gains the ability to operate on optional values, to facilitate formulation of command-line flags. When applied to two non-optional operands, the result is a non-optional String. However, if either operand has an optional type, then the concatenation has type String?, and the runtime result is None if either operand is None (which is then replaced with the empty string).

**Comment by jdidion** [comment]
@mlin you are correct, Sorry @markjschreiber. I misremembered that part of the spec. We should probably consider amending that section to say "within an interpolation, any expression that operates on a non-optional value may operate on an optional value; if the optional value is `None`, the expression is replaced with `None`, otherwise the optional value is cast to a non-optional value."

**Comment by markjschreiber** [comment]
So does that mean the best practices workflow should fail if the optional value of an Int used in the calculation in the string interpolation? Or should miniwdl produce an empty string?

**Comment by jdidion** [comment]
It should fail. Unfortunately we've found several such issues in "best-practice" workflows.

**Comment by jdidion** [comment]
https://github.com/openwdl/wdl/issues/481

**Comment by markjschreiber** [comment]
It is useful to somehow be able to construct a command line flag when a value is set and not when it's not. I don't know if abusing optionals and interpolation is the right way to do it but it's good to be able to do it.

**Comment by mlin** [comment]
@markjschreiber just to clarify, miniwdl *does* support the optional interpolation feature described in the spec language I quoted above. The stumbling block in the gatk4-rna-best-practices is specifically the attempt to do math on an optional value, something that's typically illegal, inside the interpolation. The spec carves out a special exception inside interpolations only for string concatenation, not math or other compound expressions.

**Comment by markjschreiber** [comment]
Is it possible to do something like the following pseudo code:
```
Int x
Int? maybeY
Int? maybeSum

maybeSum = (x + y)
```

and then use `maybeSum` in the interpolation?

**Comment by jdidion** [comment]
That would definitely error.

You can do

```
if (defined(maybeY)) {
  Int maybeSum = x + maybeY
}
```

We could also look at simplifying the use of optionals in expressions. For example, in fp you'd do something like `foo.unwrap()` where foo is optional and an error is raised if the value is `None`. In WDL we could add an `unwrap()` function or even a new syntax eg `foo!`.

**Comment by jdidion** [comment]
@mlin funny enough, this just came up. I wanted to do something like this and thought it would be legal:

```
version development

task foo {
  input {
    Map[String, String] m
  }

  String val = "~{m['foo']}"
  String arg = if val != "" then "--arg ~{val}" else ""

  command <<<
    mycmd ~{arg}
  >>>
}
```

According to the current version of the spec, that should work: https://github.com/openwdl/wdl/blob/5dcab3097af1753df7ebbc014cb0207d721be5b8/versions/development/SPEC.md#expression-placeholder-coercion

However, at the time this issue was originally opened, the language in the development version of the spec was different: https://github.com/openwdl/wdl/blob/5dcab3097af1753df7ebbc014cb0207d721be5b8/versions/development/SPEC.md#expression-placeholder-coercion

So we have a case where the implementation was developed based on an older version of the spec and is now contradicted by a newer version of the spec, because `development` is a moving target. I am not sure how to handle that situation.

The older language is also present in the version 1.1 spec: https://github.com/openwdl/wdl/blob/5dcab3097af1753df7ebbc014cb0207d721be5b8/versions/development/SPEC.md#expression-placeholder-coercion

However, the newer language is present in the version 1.2 spec: https://github.com/openwdl/wdl/blob/5dcab3097af1753df7ebbc014cb0207d721be5b8/versions/development/SPEC.md#expression-placeholder-coercion

So it seems that, at the very least, the above WDL should work if I replace `version development` with `version 1.2`, but it does not.
